openhtf/util/test.py

```diff
@@ -0,0 +1,387 @@
+    # you're doing (http://imgur.com/iwBCmQe).
+    self.plug_manager = plugs.PlugManager()
+    self.iterator = iterator
+    self.mock_plugs = mock_plugs
+    self.last_result = None
+
+  def _handle_phase(self, phase):
+    """Handle execution of a single test phase."""
+    # Make sure we initialize any plugs, this will ignore any that have already
+    # been initialized.
+    self.plug_manager.InitializePlugs(plug.cls for plug in phase.plugs if
+                                      plug.cls not in self.mock_plugs)
+    for plug_type, plug_value in self.mock_plugs.iteritems():
+      self.plug_manager.OverridePlug(plug_type, plug_value)
+
+    # Cobble together a fake phase data to pass to the test phase.  We use the
+    # root logger as a logger, our stub plug manager, and a dummy test record
+    # that has None for dut_id and station_id.
+    phasedata = phase_data.PhaseData(logging.getLogger(), self.plug_manager,
+                                     test_record.TestRecord(None, None))
+    phase_record = test_record.PhaseRecord(phase.name, phase.code_info)
+
+    # Actually execute the phase, saving the result in our return value.
+    with phasedata.RecordPhaseTiming(phase, phase_record):
+      try:
+        phase_record.result = phase_executor.PhaseOutcome(phase(phasedata))
+      except Exception as exc:
+        logging.exception('Exception executing phase %s', phase.name)
+        phase_record.result = phase_executor.PhaseOutcome(exc)
+
+    return phase_record
+
+  def _handle_test(self, test):
+    # Make sure we inject our mock plug instances.
+    for plug_type, plug_value in self.mock_plugs.iteritems():
+      self.plug_manager.OverridePlug(plug_type, plug_value)
+
+    # We'll need a place to stash the resulting TestRecord.
+    record_saver = RecordSaver()
+    test.AddOutputCallbacks(
+        lambda record: setattr(record_saver, 'record', record))
+
+    # Disable the station_api for unit tests.
+    test.Configure(http_port=None)
+
+    # Mock the PlugManager to use ours instead, and execute the test.
+    with mock.patch('openhtf.plugs.PlugManager', new=lambda: self.plug_manager):
+      test.Execute(test_start=lambda: 'TestDutId')
+
+    return record_saver.record
+
+  def next(self):
+    phase_or_test = self.iterator.send(self.last_result)
+    if isinstance(phase_or_test, openhtf.Test):
+      self.last_result = self._handle_test(phase_or_test)
+    elif not callable(phase_or_test):
+      raise InvalidTestError(
+          'methods decorated with patch_plugs must yield Test instances or '
+          'individual test phases', phase_or_test)
+    else:
+      self.last_result = self._handle_phase(
+          openhtf.PhaseInfo.WrapOrCopy(phase_or_test))
+    return phase_or_test, self.last_result
+
+
+def yields_phases(func):
+  """Alias for patch_plugs with no plugs patched."""
+  return patch_plugs()(func)
+
+
+def patch_plugs(**mock_plugs):
```

Comment:
Maybe consider `mock_plugs(**mocked_plugs)`? I was a little mislead by the use of "patch" at first which to me suggested something more open-ended. But maybe that's just me.